src/WorkerChannel.ts

```diff
@@ -33,32 +29,32 @@ export class WorkerChannel {
         });
     }
 
-    /**
-     * Register a patching function to be run before User Function is executed.
-     * Hook should return a patched version of User Function.
-     */
-    public registerBeforeInvocationRequest(beforeCb: InvocationRequestBefore): void {
-        this._invocationRequestBefore.push(beforeCb);
-    }
-
-    /**
-     * Register a function to be run after User Function resolves.
-     */
-    public registerAfterInvocationRequest(afterCb: InvocationRequestAfter): void {
-        this._invocationRequestAfter.push(afterCb);
+    public registerHook(hookName: string, callback: HookCallback): Disposable {
+        const hooks = this.getHooks(hookName);
+        hooks.push(callback);
+        return new Disposable(() => {
+            const index = hooks.indexOf(callback);
+            if (index > -1) {
+                hooks.splice(index, 1);
+            }
+        });
     }
 
-    public runInvocationRequestBefore(context: Context, userFunction: Function): Function {
-        let wrappedFunction = userFunction;
-        for (const before of this._invocationRequestBefore) {
-            wrappedFunction = before(context, wrappedFunction);
+    public async executeHooks(hookName: string, context: {}): Promise<void> {
+        const callbacks = this.getHooks(hookName);
+        for (const callback of callbacks) {
+            await callback(context);
         }
-        return wrappedFunction;
     }
 
-    public runInvocationRequestAfter(context: Context) {
-        for (const after of this._invocationRequestAfter) {
-            after(context);
+    private getHooks(hookName: string): HookCallback[] {
+        switch (hookName) {
+            case 'preInvocation':
+                return this._preInvocationHooks;

```

Comment:
## package split

This is a list of bullet points I see for the "@azure/functions-worker" package:
- shipped with the worker (aka built-in module. _not_ listed in user's package.json)
- slow to update
- forced on the user when it does update
- backwards compatability is paramount
- barebones API, very unopinionated

_If_ we add a new programming model package down the line, this would be its bullet points:
- shipped with the user's app (aka listed in the user's package.json dependencies)
- fast to update
- user decides when to take the update
- backwards compatability is nice, but we can easily do a new major version
- fancy api with lots of nice intellisense, can be opinionated

I do like your "separate function" suggestion, but I think that matches the second set of bullet points better. I actually already had it in my prototype [here](https://github.com/ejizba/func-nodejs-prototype/blob/main/types/index.d.ts#L57). For example, say we added a new hook "beforeDinner" at some point and a user tries to use that in their app. If they're on the wrong version of the host/worker in Azure for some reason, the existing code will throw an error "Unrecognized hook 'beforeDinner'", whereas the suggested code would throw a null ref exception because "registerBeforeDinnerHook" doesn't exist on the worker api. I think the first error is better "barebones"/"unopinionated" behavior and I think null ref errors should be avoided.

Reply:
This makes a lot of sense, thanks!